README.md
[![Latest Stable Version](https://poser.pugx.org/pbxg33k/pbxg33k-traits/v/stable)](https://packagist.org/packages/pbxg33k/pbxg33k-traits) [![Total Downloads](https://poser.pugx.org/pbxg33k/pbxg33k-traits/downloads)](https://packagist.org/packages/pbxg33k/pbxg33k-traits) [![Latest Unstable Version](https://poser.pugx.org/pbxg33k/pbxg33k-traits/v/unstable)](https://packagist.org/packages/pbxg33k/pbxg33k-traits) [![License](https://poser.pugx.org/pbxg33k/pbxg33k-traits/license)](https://packagist.org/packages/pbxg33k/pbxg33k-traits) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/PBXg33k/php-traits/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/PBXg33k/php-traits/?branch=master) [![Code Coverage](https://scrutinizer-ci.com/g/PBXg33k/php-traits/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/PBXg33k/php-traits/?branch=master)

# PBXg33k's PHP Traits

This repository is a collection of traits to make life with PHP easier

## Installation

Add the package-skeleton package to your `composer.json` file.

``` json
{
    "require": {
        "pbxg33k/pbxg33k-traits": "1.0.*"
    }
}
```

Or via the command line in the root of your project's installation.

``` bash
$ composer require "pbxg33k/pbxg33k-traits*"
```

## Traits


- **HydratableTrait** Allows you to easily hydrate classes from arrays. An example is importing data from external APIs 
- **ReflectionTrait** Allows you to do extra things with Reflection (ie: get property class from @var block)
- **PropertyTrait** Set property values without worrying about property visibility or setters

## Usage
Click [here](http://php.net/manual/en/language.oop5.traits.php) to read about using traits on PHP's own manual.

### HydratableTrait ###
```php
class Foo 
{
    use Pbxg33k\Traits\HydratableTrait;
    // Rest of your class

    // Example property, imagine it has proper getter/setter
    protected $randomProperty;
}

// Somewhere else in code
$foo = new Foo();
$foo->hydrateClass(['randomProperty' => 'value']);

var_dump($foo->getRandomProperty()); // "value"
```

HydratableTrait trait allows you to hydrate your class properties easily by passing an array to hydrateClass().
This trait will automagically assign matching keys to properties and instantiate supported classes.


### ReflectionTrait ###
```php
Class Foo
{
    /**
     * @var DateTime
     */
    property $dateTime;
}

Class BlackMagic
{
    use Pbxg33k\Traits\ReflectionTrait;

    public static function getPropertyClass($class, $property)
    {
        // Wrapped following method because it's protected
        return $this->getClassFromClassProperty($class, $property);
    }
}

BlackMagic::getPropertyClass(Foo::class, 'dateTime'); // returns "DateTime"

```

## Contributing

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request

## Changelog

Please see [CHANGELOG.md](CHANGELOG.md)


## License

The MIT License (MIT)
Copyright (c) 2016 Oguzhan Uysal.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## v1.0.0
### Added
- HydratableTrait trait `Pbxg33k\Traits\HydratableTrait`
- ReflectionTrait trait `Pbxg33k\Traits\ReflectionTrait`
- PropertyTrait trait `Pbxg33k\Traits\PropertyTrait`

src/PropertyTrait.php
<?php
/**
 * Created by PhpStorm.
 * User: PBX_g33k
 * Date: 24-May-16
 * Time: 13:02
 */

namespace Pbxg33k\Traits;

/**
 * Class PropertyTrait
 *
 * This trait adds helpers for getters and setters by automatically trying
 * to set properties using various methods while respecting property visibility.
 *
 * @author  Oguzhan uysal <development@oguzhanuysal.eu>
 * @package Pbxg33k\Traits
 */
trait PropertyTrait
{
    /**
     * Sets the property value for the object
     *
     * @param string $key
     * @param mixed $value
     * @param bool $override
     * @param bool $isCollection Boolean to indicate Doctrine Collections
     * @return Object
     * @throws \Exception if setting value has failed
     */
    public function setPropertyValue($key, $value, $override = false, $isCollection = false)
    {
        // Replace empty strings with null
        $value = ($value === "") ? null : $value;

        // Convert key to method names
        $methodName = $this->getMethodName($key);

        if (!$override && $this->getPropertyValue($key) === $value) {
            return $this;
        }

        $setMethod = ($isCollection) ? 'add' : 'set';
        if (method_exists($this, $setMethod . $methodName)) {
            return $this->{$setMethod . $methodName}($value);
        } else {
            throw new \Exception(sprintf(
                'Unable to set value, method not found: %s%s in class: %s',
                $setMethod,
                $methodName,
                static::class
            ));
        }
    }

    /**
     * Get property value
     *
     * @param $key
     * @return mixed
     */
    public function getPropertyValue($key) {
        $methodName = 'get' . $this->getMethodName($key);
        if(method_exists($this, $methodName)) {
            return $this->{$methodName}();
        }
        return false;
    }

    /**
     * Converts snake_case to CamelCase to convert property names to getters
     *
     * @param  string $propertyKey
     * @return string
     */
    protected function getMethodName($propertyKey)
    {
        return ucfirst(preg_replace_callback('/_([a-z])/', function ($match) {
            return strtoupper($match[1]);
        }, $propertyKey));
    }
}

src/ReflectionTrait.php
<?php
namespace Pbxg33k\Traits;

/**
 * Class ReflectionTrait
 *
 * This trait adds methods to retrieve properties easily using PHP Reflection
 *
 * @author  Oguzhan Uysal <development@oguzhanuysal.eu>
 * @package Pbxg33k\Traits
 */
trait ReflectionTrait
{
    /**
     * Returns Class from the DockBlock above a property
     * @param $class
     * @param $property
     *
     * @throws \Exception
     */
    public function getClassFromClassProperty($class, $property)
    {
        if(!class_exists($class)) {
            throw new \Exception($class. ' not found or does not exist');
        }
        
        if(!property_exists($class, $property)) {
            throw new \Exception($class . ' has no property with the name ' . $property);
        }
    }
    
    /**
     * Tries to get the correct class name from the given docBlock for Reflection
     *
     * @param string $comment the docblock
     * @param bool $includeNamespaces
     * @param null|\ReflectionClass $reflectionClass
     *
     * @return bool|string
     */
    public static function getClassFromDocComment($comment, $includeNamespaces = true, $reflectionClass = null)
    {
        if (preg_match('~\@var[\s]+([A-Za-z0-9\\\\]+)~', $comment, $matches)) {
            if ($includeNamespaces) {
                if ($reflectionClass instanceof \ReflectionClass && !in_array($matches[1], HydratableTrait::$nonObjectTypes)) {
                    return ($reflectionClass->getNamespaceName()) ? sprintf('\%s\%s', $reflectionClass->getNamespaceName(), $matches[1]) :  sprintf('\%s', $matches[1]);
                }
                return $matches[1];
            }
            return join('', array_slice(explode('\\', $matches[1]), -1));
        }

        return false;
    }
}

src/HydratableTrait.php
<?php
/**
 * Created by PhpStorm.
 * User: PBX_g33k
 * Date: 24-May-16
 * Time: 00:07
 */

namespace Pbxg33k\Traits;

/**
 * Class HydratableTrait
 *
 * This trait allows you to hydrate a class/object by passing an array or stdObj and let it hydrate itself by
 * calling either of the following methods:
 *  - hydrateClass($rawData)
 *
 * @author  Oguzhan Uysal <development@oguzhanuysal.eu>
 * @package Pbxg33k\Traits
 */
trait HydratableTrait
{
    use PropertyTrait;
    use ReflectionTrait;

    /**
     * List of types which are not used as objects
     *
     * @var array
     */
    public static $nonObjectTypes = ['string', 'int', 'integer', 'bool', 'boolean', 'array', 'float', 'mixed', 'null'];

    /**
     * List of classes which will take string arguments in constructor
     *
     * @var array
     */
    protected $giveDataInConstructor = ['\DateTime'];

    /**
     * Object constructor arguments to be passed when creating an object during conversion
     *
     * @var mixed
     */
    protected $objectConstructorArguments;

    /**
     * Converts a stdClass to models loaded in current context
     *
     * This method iterates over the passed $class
     * For each key, it looks for a setter and type.
     * If the value is an object, it initializes the object and assignes the initialized object.
     *
     * @param  object|array $class       class
     * @param  boolean      $failOnError Throw Exception if any error(s) occur
     *
     * @return object
     *
     * @throws \Exception if hydration failes AND $failOnError is true
     */
    public function hydrateClass($class, $failOnError = false)
    {
        $reflection = new \ReflectionClass($this);

        // Iterate over $class for properties
        foreach ($class as $itemKey => $itemValue) {
            // Convert key to a propertyname in $this
            try {
                $this->hydrateProperty($itemKey, $itemValue, $reflection);
            } catch (\Exception $e) {
                if ($failOnError) {
                    throw $e;
                }
            }
        }

        return $this;
    }

    /**
     * Hydrates property with value.
     * Value can be anything. If the property is supposed to be a Class of anykind we will try to instantiate it
     * and assign the class to the property
     *
     * @param                  $key
     * @param                  $value
     * @param \ReflectionClass $reflection
     *
     * @throws \Exception
     */
    protected function hydrateProperty($key, $value, \ReflectionClass $reflection)
    {
        $propertyName = $this->resolvePropertyName($key);

        try {
            // Check if property exists and assign a ReflectionProperty class to $reflectionProperty
            $reflectionProperty = $reflection->getProperty($propertyName);
            // Get the expected property class from the property's DocBlock
            $propertyClassName = ReflectionTrait::getClassFromDocComment($reflectionProperty->getDocComment(), true, $reflection);
            // Set argument for constructor (if any), in case we're dealing with an object (IE: DateTime)
            $this->objectConstructorArguments = (in_array($propertyClassName, $this->giveDataInConstructor)) ? $value : null;

            if (!in_array($propertyClassName, self::$nonObjectTypes) && class_exists($propertyClassName)) {
                $object = new $propertyClassName($this->objectConstructorArguments);
                $this->checkObjectForErrors($object, true);

                if (method_exists($object, 'hydrateClass') && $this->isHydratableValue($value)) {
                    $object->hydrateClass($value);
                }
                $value = $object;
            }

            $this->setPropertyValue($propertyName, $value, true);
        } catch (\Exception $e) {
            throw $e;
        }
    }

    /**
     * Resolves and returns propertyname
     *
     * @param $key
     *
     * @return mixed|string
     */
    private function resolvePropertyName($key)
    {
        $propertyName = lcfirst(str_replace(' ', '', ucwords(str_replace('_', ' ', $key))));

        return (property_exists($this, $propertyName)) ? $propertyName :
            (property_exists($this, lcfirst($propertyName)) ? lcfirst($propertyName) :
                preg_replace_callback('/([A-Z])/', function($match) {
                    return strtolower('_' . $match[1]);
                }, lcfirst($propertyName))
            );
    }

    /**
     * Checks if the value can be hydrated for iteration
     *
     * @param $value
     *
     * @return bool
     */
    private function isHydratableValue($value)
    {
        return (is_array($value) || is_object($value));
    }

    /**
     * Checks (and fixes) objects against known errors
     *
     * @param Object &$object
     * @param bool   $fix Fix errors
     *
     * @return void
     *
     * @throws \Exception
     */
    private function checkObjectForErrors(&$object, $fix = false)
    {
        if ($object instanceof \DateTime) {
            if ($this->objectConstructorArguments == null) {
                $object = null;
            }
        }
    }
}